### PR TITLE
Improve runtime linkage

### PR DESCRIPTION
**Pull request by akiramenai**
The PR consists of 3 patches. The first one addresses issue of compilation time increase due to early runtime linkage:
```
Split runtime in two parts: runtime and stdlib. Stdlib is the part to be
linked and internalized in the beginning of opt pipeline. Then global
DCE removes the unused parts. So small contracts doesn't blow up and
affect compilation time. Runtime is linked in the beginning of llc
pipeline.

This solution replaces previously used internalize-early,
internalize-late function attributes.
```
The remaining two, remove local changes in old pm pipeline.

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                               -0.364 ║
║ Best                                3.478 ║
║ Worst                             -13.494 ║
║ Total                              -0.317 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                3.125 ║
║ Worst                             -12.727 ║
║ Total                              -0.012 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.055 ║
║ Best                                0.786 ║
║ Worst                             -13.494 ║
║ Total                              -0.279 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -0.541 ║
║ Best                                0.047 ║
║ Worst                              -2.113 ║
║ Total                              -0.247 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                               -0.364 ║
║ Best                                3.478 ║
║ Worst                             -13.494 ║
║ Total                              -0.317 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                3.125 ║
║ Worst                             -12.727 ║
║ Total                              -0.012 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.055 ║
║ Best                                0.786 ║
║ Worst                             -13.494 ║
║ Total                              -0.279 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -0.541 ║
║ Best                                0.047 ║
║ Worst                              -2.113 ║
║ Total                              -0.247 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMLinkRuntime.cpp [hunk 1]
```diff
@@ -38,23 +39,15 @@ void initializeSyncVMLinkRuntimePass(PassRegistry &);
...
 namespace {
-/// Link the runtime library into the module.
+/// Link std and runtime libraries into the module.
 /// At the moment front ends work only with single source programs.
 struct SyncVMLinkRuntime : public ModulePass {
 public:
   static char ID;
-  SyncVMLinkRuntime(bool OnlyInternalize = false)
-      : ModulePass(ID), OnlyInternalize(OnlyInternalize) {
+  SyncVMLinkRuntime(bool IsRuntimeLinage = false)
```
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMLinkRuntime.cpp]: nit: typo.

File: llvm/lib/Target/SyncVM/SyncVMLinkRuntime.cpp [hunk 2]
```diff
@@ -68,83 +61,59 @@ struct SyncVMLinkRuntime : public ModulePass {
   }
 
 private:
-  bool OnlyInternalize;
+  bool IsRuntimeLinage;
```
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMLinkRuntime.cpp]: nit: typo here too.

**Review by lialan — APPROVED** [review]
Just a typo, otherwise LGTM.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Let's rename 'ergs' => 'gas' in this PR.

**Comment by akiramenai** [comment]
@hedgar2017 did it in a separate one as renaming has nothing to do with linkage.

**Review by hedgar2017 — APPROVED** [review]

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.019 ║
║ Best                                2.326 ║
║ Worst                               0.000 ║
║ Total                               0.002 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by akiramenai** [comment]
All 18 failures are due to `context.set_gas_per_pubdata`, so I'm pushing without double check.

**Pull request merged**